core/embed/extmod/modtrezorcrypto/modtrezorcrypto-hmac.h

```diff
@@ -0,0 +1,166 @@
+#define SHA512 512
+
+/// package: trezorcrypto.__init__
+
+/// class hmac:
+///     """
+///     HMAC context.
+///     """
+///     SHA256: int
+///     SHA512: int
+typedef struct _mp_obj_Hmac_t {
+  mp_obj_base_t base;
+  union {
+    HMAC_SHA256_CTX ctx256;
+    HMAC_SHA512_CTX ctx512;
+  };
+  uint32_t hashtype;
+} mp_obj_Hmac_t;
+
+STATIC mp_obj_t mod_trezorcrypto_Hmac_update(mp_obj_t self, mp_obj_t data);
+
+/// def __init__(
+///     self,
+///     hashtype: int,
+///     key: bytes,
+///     message: bytes = None
+/// ) -> None:
+///     """
+///     Create a HMAC context.
+///     """
+STATIC mp_obj_t mod_trezorcrypto_Hmac_make_new(const mp_obj_type_t *type,
+                                               size_t n_args, size_t n_kw,
+                                               const mp_obj_t *args) {
+  mp_arg_check_num(n_args, n_kw, 2, 3, false);
+  mp_obj_Hmac_t *o = m_new_obj_with_finaliser(mp_obj_Hmac_t);
+  o->base.type = type;
+
+  mp_buffer_info_t key = {0};
+  mp_get_buffer_raise(args[1], &key, MP_BUFFER_READ);
+  mp_buffer_info_t msg = {0};
+  // constructor called with message as third parameter
+  if (n_args > 2) {
+    mp_get_buffer_raise(args[2], &msg, MP_BUFFER_READ);
+  }
+
+  if (key.len == 0) {
+    key.buf = "";
+  }
+
+  o->hashtype = trezor_obj_get_uint(args[0]);
+  if (o->hashtype == SHA256) {
+    hmac_sha256_Init(&(o->ctx256), key.buf, key.len);
+    if (n_args > 2) {
+      hmac_sha256_Update(&(o->ctx256), msg.buf, msg.len);
+    }
+  } else if (o->hashtype == SHA512) {
+    hmac_sha512_Init(&(o->ctx512), key.buf, key.len);
+    if (n_args > 2) {
+      hmac_sha512_Update(&(o->ctx512), msg.buf, msg.len);
+    }
+  } else {
+    mp_raise_ValueError("Invalid hashtype");
+  }
+  return MP_OBJ_FROM_PTR(o);
+}
+
+/// def update(self, message: bytes) -> None:
+///     """
+///     Update a HMAC context.
+///     """
+STATIC mp_obj_t mod_trezorcrypto_Hmac_update(mp_obj_t self, mp_obj_t message) {
+  mp_obj_Hmac_t *o = MP_OBJ_TO_PTR(self);
+  mp_buffer_info_t msg = {0};
+  mp_get_buffer_raise(message, &msg, MP_BUFFER_READ);
+  if (o->hashtype == SHA256) {
+    hmac_sha256_Update(&(o->ctx256), msg.buf, msg.len);
+  } else if (o->hashtype == SHA512) {
+    hmac_sha512_Update(&(o->ctx512), msg.buf, msg.len);
+  } else {
+    mp_raise_ValueError("Invalid hashtype");
+  }
+  return mp_const_none;
+}
+STATIC MP_DEFINE_CONST_FUN_OBJ_2(mod_trezorcrypto_Hmac_update_obj,
+                                 mod_trezorcrypto_Hmac_update);
+
+/// def digest(self) -> bytes:
+///     """
+///     Retrieve the digest key.
+///     """
+STATIC mp_obj_t mod_trezorcrypto_Hmac_digest(mp_obj_t self) {
+  mp_obj_Hmac_t *o = MP_OBJ_TO_PTR(self);
+  if (o->hashtype == SHA256) {
+    HMAC_SHA256_CTX ctx = {0};
+    memcpy(&ctx, &(o->ctx256), sizeof(HMAC_SHA256_CTX));
+    uint8_t out[SHA256_DIGEST_LENGTH] = {0};
+    hmac_sha256_Final(&ctx, out);
+    memzero(&ctx, sizeof(HMAC_SHA256_CTX));
+    return mp_obj_new_bytes(out, sizeof(out));
```

Comment:
```suggestion
    vstr_t vstr = {0};
    vstr_init_len(&vstr, SHA256_DIGEST_LENGTH);
    hmac_sha256_Final(&ctx, (uint8_t *)vstr.buf);
    memzero(&ctx, sizeof(HMAC_SHA256_CTX));
    return mp_obj_new_str_from_vstr(&mp_type_bytes, &vstr);
```

Untested. Note that the same applies to the SHA-512 branch below.

BTW `hmac_sha256_Final()` already calls `memzero` on the context, so technically it's not needed here.

Reply:
This seems fine, but we use `mp_obj_new_bytes` all over the modtrezorcrypto. I prefer consistency. If we think this should be changed (and I am not opposed), let's address this in another PR where we replace all calls to `mp_obj_new_bytes` with `mp_obj_new_str_from_vstr`.

Reply:
There is nothing wrong with using `mp_obj_new_bytes()` as long as we clean up the stack, which we don't do here.

The `aes`, `bip32`, `chacha20poly1305`, `random` and `shamir` modules use `vstr`, but you are right that the hash-related modules and some other important ones don't. I will create a PR for that.